### PR TITLE
adding view_trie to state-viewer

### PR DESCRIPTION
**Pull request by mm-near**
Adding an option to view the trie nodes using state-viewer.

After this PR, you can run commands like:
```
neard --home ~/.near view-state  view-trie --hash HKhtLNsdzjJdN63AD2gqWVyRn6ASopnxmqv6weQoiqFy --shard-id 0 --max-depth 5
```
To see the detailed info about the trie nodes.

More info in: https://nearinc.atlassian.net/wiki/spaces/EAP/pages/136609793/Using+state-viewer

### CODE REVIEW

File: core/store/src/trie/mod.rs [hunk 1]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+            Ok(Some((_, raw_node))) => {
+                match raw_node.node {
+                    RawTrieNode::Leaf(key, value_length, value_hash) => {
+                        let (slice, _) = NibbleSlice::from_encoded(key.as_slice());
+                        prefix.extend(slice.iter());
+                        writeln!(
+                            f,
+                            "{}Leaf {:?} size:{} child_hash:{} child_prefix:{}",
+                            spaces,
+                            slice,
```
Comment (mina86) [inline core/store/src/trie/mod.rs:642]: Should we also use nibbles_to_string here?  Perhaps we should add Display implementation to NibbleSlice?

File: core/store/src/trie/mod.rs [hunk 2]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+                            f,
+                            "{}Branch Value:{:?} prefix:{}",
+                            spaces,
+                            optional_value,
+                            self.nibbles_to_string(prefix)
+                        )?;
+                        for (idx, child) in children.iter().enumerate() {
+                            if let Some(child) = child {
+                                writeln!(f, "{}{:01x}->", spaces, idx)?;
+                                spaces.push(' ');
```
Comment (mina86) [inline core/store/src/trie/mod.rs:660]: Perhaps change to indenting by two spaces? This way the cases in branch would have their own indentation level.
```suggestion
                                writeln!(f, "{} {:01x}->", spaces, idx)?;
                                spaces.push_str("  ");
```

File: core/store/src/trie/mod.rs [hunk 3]
```diff
@@ -586,6 +586,130 @@ impl Trie {
         Ok(())
     }
 
+    // Prints the trie nodes starting from hash, up to max_depth depth.
+    pub fn print_recursive(&self, hash: &CryptoHash, max_depth: u32) -> String {
+        let mut buf = String::new();
+        let mut prefix: Vec<u8> = Vec::new();
+        self.print_recursive_internal(&mut buf, hash, max_depth, &mut "".to_string(), &mut prefix)
```
Comment (mina86) [inline core/store/src/trie/mod.rs:593]: Alternatively `&mut String::with_capacity(max_depth * 2)` as the third argument.

File: core/store/src/trie/mod.rs [hunk 4]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+        let mut result = String::new();
+        for i in (0..prefix.len()).step_by(2) {
+            if i + 1 < prefix.len() {
+                let chr: char = ((prefix[i] * 16) + prefix[i + 1]).into();
+                result.push_str(&format!("{}", chr.escape_default()));
+            }
+        }
+        if prefix.len() % 2 == 1 {
+            result.push_str(&format!(" + {}", prefix[prefix.len() - 1]));
+        }
```
Comment (mina86) [inline core/store/src/trie/mod.rs:609]:
```suggestion
        for chunk in prefix.chunks(2) {
            if chunk.len() == 2 {
                let chr: char = (chunk[0] * 16 + chunk[1]).into();
                write!(&mut result, chr.escape_default());
            } else {
                // Final, sole nibble
                write!(&mut result, " + {:x}", chunk[0]);
            }
        }
```

File: core/store/src/trie/mod.rs [hunk 5]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+        if max_depth == 0 {
+            writeln!(
+                f,
+                "{}max depth reached hash:{} prefix:{}",
+                spaces,
+                hash,
+                self.nibbles_to_string(prefix)
+            )?;
+            return Ok(());
+        }
```
Comment (mina86) [inline core/store/src/trie/mod.rs:630]: This feels like it would be a noisy message.  The hash and prefix had already been printed by the caller and we can see that max depth was reached by the fact that nothing deeper is printed. ;)

File: core/store/src/trie/mod.rs [hunk 6]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+                                prefix.push(idx as u8);
+                                self.print_recursive_internal(
+                                    f,
+                                    child,
+                                    max_depth - 1,
+                                    spaces,
+                                    prefix,
+                                )?;
+                                prefix.pop();
+                                spaces.remove(spaces.len() - 1);
```
Comment (mina86) [inline core/store/src/trie/mod.rs:670]:
```suggestion
                                spaces.truncate(spaces.len() - 2);
```

File: core/store/src/trie/mod.rs [hunk 7]
```diff
@@ -595,7 +719,12 @@ impl Trie {
         }
         let bytes = self.storage.retrieve_raw_bytes(hash)?;
         let node = RawTrieNodeWithSize::decode(&bytes).map_err(|err| {
-            StorageError::StorageInconsistentState(format!("Failed to decode node {hash}: {err}"))
+            StorageError::NotANode(NotANodeError {
+                hash: hash.clone(),
+                len: bytes.len(),
+                data_prefix: format!("{:?}", &bytes[..std::cmp::min(10, bytes.len())]),
+                error: format!("Failed to decode node {hash}.: {err}"),
```
Comment (mina86) [inline core/store/src/trie/mod.rs:726]: Perhaps instead just `err: err` and then Display / Debug can format it however it wants?

File: core/store/src/trie/mod.rs [hunk 8]
```diff
@@ -595,7 +719,12 @@ impl Trie {
         }
         let bytes = self.storage.retrieve_raw_bytes(hash)?;
         let node = RawTrieNodeWithSize::decode(&bytes).map_err(|err| {
-            StorageError::StorageInconsistentState(format!("Failed to decode node {hash}: {err}"))
+            StorageError::NotANode(NotANodeError {
+                hash: hash.clone(),
+                len: bytes.len(),
+                data_prefix: format!("{:?}", &bytes[..std::cmp::min(10, bytes.len())]),
```
Comment (mina86) [inline core/store/src/trie/mod.rs:725]: This screams to me SmallVec.  We might want to start using it.

File: core/store/src/trie/mod.rs [hunk 9]
```diff
@@ -586,6 +586,130 @@ impl Trie {
         Ok(())
     }
 
+    // Prints the trie nodes starting from hash, up to max_depth depth.
+    pub fn print_recursive(&self, hash: &CryptoHash, max_depth: u32) -> String {
```
Comment (mina86) [inline core/store/src/trie/mod.rs:590]: Perhaps pass output as argument so we don’t need to collect it to String first?
```suggestion
    pub fn print_recursive(&self, out: &mut dyn std::io::Write, hash: &CryptoHash, max_depth: u32) -> String {
```

File: tools/state-viewer/src/cli.rs [hunk 10]
```diff
@@ -447,3 +451,22 @@ impl ApplyReceiptCmd {
...
+pub struct ViewTrieCmd {
+    #[clap(long)]
+    hash: String,
+    #[clap(long)]
+    shard_id: u32,
+    #[clap(long)]
+    shard_version: u32,
+    #[clap(long)]
+    max_depth: u32,
+}
```
Comment (mina86) [inline tools/state-viewer/src/cli.rs:465]: It would be useful to also have `prefix: Option<String>` so that I can tell `view-trie ... --max-depth=5 --prefix=1234ABCD5678` and it would start printing five levels deep from 1234ABCD7678 prefix.

File: core/primitives/src/errors.rs [hunk 11]
```diff
@@ -84,6 +92,9 @@ pub enum StorageError {
     /// panic in every place that produces this error.
     /// We can check if db is corrupted by verifying everything in the state trie.
     StorageInconsistentState(String),
+    /// Error thrown when a given value cannot be parsed as a TrieNode.
+    /// This probably means that this is a 'value' field (as we store both in the same colState).
+    NotANode(NotANodeError),
```
Comment (matklad) [inline core/primitives/src/errors.rs:97]: `NotANode` is a kind of `StorageInconsistentState`. If you want to have it well-typed, then I sugest adding

```
enum StorageInconsistentState { Other(String), NotANodeError(NotANodeError) }
```

File: tools/state-viewer/src/cli.rs (hunk 10 above)
Comment (mm-near) [reply tools/state-viewer/src/cli.rs:465]: yeah - I plan to do it in a separate PR (having the prefix as you've mentioned and also an ability to pass the account_id directly)

File: core/store/src/trie/mod.rs [hunk 12]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+        let mut buf = String::new();
+        let mut prefix: Vec<u8> = Vec::new();
+        self.print_recursive_internal(&mut buf, hash, max_depth, &mut "".to_string(), &mut prefix)
+            .expect("printing failed");
+        buf
+    }
+
+    // Converts the list of Nibbles to a readable string.
+    fn nibbles_to_string(&self, prefix: &Vec<u8>) -> String {
```
[the remark this answers is not in the document]
Comment (mm-near) [reply core/store/src/trie/mod.rs:599]: done.

File: core/store/src/trie/mod.rs (hunk 4 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:609]: very nice -thanks.

File: core/store/src/trie/mod.rs [hunk 13]
```diff
@@ -586,6 +586,130 @@ impl Trie {
...
+                        spaces.push(' ');
+                        prefix.extend(slice.iter());
+                        self.print_recursive_internal(f, &child, max_depth - 1, spaces, prefix)?;
+                        prefix.truncate(prefix.len() - slice.len());
+                        spaces.remove(spaces.len() - 1);
+                    }
+                };
+            }
+            Ok(None) => {
+                writeln!(f, "none")?;
```
[the remark this answers is not in the document]
Comment (mm-near) [reply core/store/src/trie/mod.rs:693]: done

File: core/store/src/trie/mod.rs (hunk 2 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:660]: good idea. done.

File: core/store/src/trie/mod.rs (hunk 6 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:670]: done.

File: core/store/src/trie/mod.rs (hunk 7 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:726]: done.

File: core/store/src/trie/mod.rs (hunk 5 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:630]: good point.

File: core/store/src/trie/mod.rs (hunk 1 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:642]: if we add nibble_to_string here, we then probably should look if we're in the 'even' or 'odd' depth (and that's why I've simply printed the slice, rather than complicate the code)

File: core/primitives/src/errors.rs (hunk 11 above)
Comment (mm-near) [reply core/primitives/src/errors.rs:97]: it is hard to say if this is the 'inconsistency'...
This is one weakness of our design - where we put two types of objects into one column (Nodes and values), without any indication on what to expect.


The alternative approach, is to get rid of this error and add another function similar to "fn retrieve_raw_node" - something like 'retrieve_raw_node_or_value" instead.

File: core/store/src/trie/mod.rs (hunk 9 above)
Comment (mm-near) [reply core/store/src/trie/mod.rs:590]: maybe I'm missing something: but how does it help?

As the caller would still have to do:
let mut buf = String::new();
print_resursive(&buf, ...)
println!("{}", buf)

right?
(or would the caller be able to do something smarter?)

Comment (mina86) [reply core/store/src/trie/mod.rs:644]: Just pass `&mut std::io::stdout().lock()` without accumulating into a string.

**Review by mina86 — APPROVED** [review]

File: core/primitives/src/errors.rs (hunk 11 above)
Comment (matklad) [reply core/primitives/src/errors.rs:97]: Hm, I think this case is pretty clear: today, the contract is that, if you call `retrieve_raw_node` and hit the case where the node fails to decode, that means 

* either a programming error
* or the database being seriously messed up and lying to us about its contents. 

So, when the caller gets storage inconsistent state, it's pretty clear what happens, and what can be done (not much). 


Now, in this PR we add one more variant, and the problem here is that now the call-site has no idea what to do with `NotANode`. It's very fuzzy whether this is a programming error, a catastrophic DB failure, an invalid challenge, or maybe just a normal outcome. 

So, yeah, I would much prefer to add 

```rust
fn retrieve_raw_node_or_value() -> Result<Result<Node, NotANode>, StorageError>
```

this won't infect every existing caller of a `StorageError`-returning function. 


---


Looking more holistically here, it seems it should never be the case that we don't know whether we expect a node or a value? The root of the trie is always a node, and, when we descend, we always know whether we are fetching the root node or a value.

In other words, I think the usage of this error in `print_recursive_internal` is, pedantically speaking, incorrect -- it's only on the first level of recursion where we don't know whether this is  node or a value, all recursive calls must treat this condition as a messed-up database. So, we rather want something like this instead: 


```
pub fn print_recursive(&self, hash: &CryptoHash, max_depth: u32) -> String {
        if let Err() = self.get_raw_trie_node(hash) {
            panic!("{hash} doesn't point to a trie node")   
        }
        self.print_recursive_internal(hash)
}
```

File: core/store/src/trie/mod.rs (hunk 8 above)
Comment (matklad) [reply core/store/src/trie/mod.rs:772]: I would disagree here -- this is an error, and for errors you generally care about small size_of, rather than few allocations (allocations only affect sad path, size_of affects all paths). So what we want here is actually the oppostie -- to box this stuff away, such that the whole error is just a single pointer. 

But either way i expect the impact to be small.

Comment (mina86) [reply core/store/src/trie/mod.rs:772]: > this is an error, and for errors you generally care about small size_of, rather than few allocations

SmallVec capable of holding 16 bytes inline is the same size as String or Vec.  We can optimise allocations without needing to sacrifice size.

And if we want to optimise size than there’s also ArrayVec.

File: core/primitives/src/errors.rs (hunk 11 above)
Comment (mm-near) [reply core/primitives/src/errors.rs]: thanks for the tips - done.
